raku-advent-2021/unix_philosophy_without_leftpad.1.md

```diff
@@ -0,0 +1,120 @@
+
+In the context of writing libraries, the Unix philosophy encourages the practice of writing micro-packages: small libraries, intentionally limited in scope, that serve exactly one purpose.  Some programming language communities have this as an explicit goal; for example, one of the leading Node.js developers [explicitly invoked the Unix philosophy](https://blog.izs.me/2013/04/unix-philosophy-and-nodejs/) in their advice to programmers:
+
+> Write modules that do one thing well. Write a new module rather than complicate an old one.
+
+This practice of writing micro packages contrasts sharply with the practice of writing [omnibus packages](https://jquery.com/) that attempt to provide a single, coherent API that aims to  solve any problem a developer might encounter.  And micro packages benefit from all the advantages that have made the Unix philosophy such good advice for 50 years.  Most notably, micro packages tend to be simple enough (and **small** enough) that you can personally inspect the code – and, if necessary, debug any issues that come up.
+
+### The downside of micro packages
+
+As this post's title probably gave away, the problem with overusing micro packages is that it can lead to what happened with left-pad.  Without rehashing [all the details](https://www.davidhaney.io/npm-left-pad-have-we-forgotten-how-to-program/), there was an 11-line JavaScript package (left-pad) that did nothing other than pad each line of a string with a specified amount of whitespace.  Yet, somehow, a huge percentage of the JavaScript ecosystem depended on this simple function – either directly or more commonly indirectly.  As a result, when the developer removed the package (in a way that couldn't happen anymore for reasons not relevant here), that same fraction of the JavaScript ecosystem fell over.  I'm not sure exactly how many builds failed, but [one source](https://www.theregister.com/2016/03/23/npm_left_pad_chaos/) estimated that over 2.4 _million_ software builds depended on left-pad every month.  So not a small number.
+
+In other wordsSo, basically, someone finally pulled out thate one domino that the entire Internet depend on:
+
+[![xkcd 2347. [A tower of blocks is shown. The upper half consists of many tiny blocks balanced on top of one another to form smaller towers, labeled:]  All modern digital infrastructure  [The blocks rest on larger blocks lower down in the image, finally on a single large block. This is balanced on top of a set of blocks on the left, and on the right, a single tiny block placed on its side. This one is labeled:]  A project some random person in Nebraska has been thanklessly maintaining since 2003.  {alt-text:} Someday ImageMagick will finally break for good and we'll have a long period of scrambling as we try to reassemble civilization from the rubble.](https://codesections.com/xkcd_2347.png)](https://xkcd.com/2347/)
+
```

Comment:
You sure you can copy that image? Or will you embed it from original source?

Reply:
xkcd is covered by a CC-BY-NC license and explicitly allows this type of use (with a link back, as I have here).  See https://xkcd.com/license.html.  I added an indication of the license though (always a good idea).  Thanks!